### DISCUSSION TITLE
JDBC driver doesn't allow PreparedStatement.getMetaData() before .executeQuery()

### DISCUSSION THREAD

**Discussion by rkennedy-mode** [duckdb/duckdb#9143] [Q&A] [answered]
I have some code that executes the following:

```java
try (var preparedStatement = connection.prepareStatement("SELECT * FROM table_name")) {
    preparedStatement.getMetaData();
}
```

The JDBC docs for [PreparedStatement.getMetaData()](https://docs.oracle.com/en/java/javase/20/docs/api/java.sql/java/sql/PreparedStatement.html#getMetaData()) indicate that this is something that **can** be allowed if preparing the statement means that the underlying database knows ahead of time what the resulting columns (names, types, etc) will be:

> Because a PreparedStatement object is precompiled, it is possible to know about the ResultSet object that it will return without having to execute it. Consequently, it is possible to invoke the method getMetaData on a PreparedStatement object rather than waiting to execute it and then invoking the ResultSet.getMetaData method on the ResultSet object that is returned. 

For example, the Apache Derby JDBC driver allows this.

In the DuckDB driver (0.8.1) this causes an exception:

```
java.sql.SQLException: Prepare and execute something first
	at org.duckdb.DuckDBPreparedStatement.getMetaData(DuckDBPreparedStatement.java:207)
	at com.mode.ryankennedy.duckflight.producer.MyFlightSqlProducer.lambda$createPreparedStatement$1(MyFlightSqlProducer.java:79)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1623)
```

I'm wondering if this is something that DuckDB itself can't do or if it's something that it could do but would need work in the driver to support.

**Comment by Mause** [comment, marked as answer]
Hey there, can you please try with 0.9.0? I'm fairly sure I already fixed this

**Reply by rkennedy-mode** [reply]
Oh, amazing! Yes, this works with 0.9.0. I hadn't noticed all of the JDBC changes in the [0.9.0 release announcement](https://duckdb.org/2023/09/26/announcing-duckdb-090.html).

Is there a general theme for the JDBC updates in 0.9.0? It looks like we discovered another (possibly) fortunate change for us in 0.9.0 via [duckdb/foundation-discussions/discussions/18](https://github.com/duckdb/foundation-discussions/discussions/18#discussioncomment-7135391).

**Reply by rkennedy-mode** [reply]
Ah, it looks like this was done as part of #8280, which is referenced from the full release notes. 🎉 

Thank you, @Mause!
